central/imagev2/datastore/mapper/datastore/datastore_impl.go

```diff
@@ -0,0 +1,208 @@
+package datastore
+
+import (
+	"context"
+
+	imageDatastore "github.com/stackrox/rox/central/image/datastore"
+	imageV2Datastore "github.com/stackrox/rox/central/imagev2/datastore"
+	"github.com/stackrox/rox/central/imagev2/datastore/mapper"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/features"
+	"github.com/stackrox/rox/pkg/images/types"
+	searchPkg "github.com/stackrox/rox/pkg/search"
+)
+
+type datastoreImpl struct {
+	imageDataStore   imageDatastore.DataStore
+	imageV2DataStore imageV2Datastore.DataStore
+
+	flattenImageData bool
+}
+
+func newDatastoreImpl(datastoreV1 imageDatastore.DataStore, datastoreV2 imageV2Datastore.DataStore) *datastoreImpl {
+	ds := &datastoreImpl{
+		imageDataStore:   datastoreV1,
+		imageV2DataStore: datastoreV2,
+
+		flattenImageData: features.FlattenImageData.Enabled(),
+	}
+	return ds
+}
+
+func (ds *datastoreImpl) SearchListImages(ctx context.Context, q *v1.Query) ([]*storage.ListImage, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.SearchListImages(ctx, q)
+	}
+
+	// Get v2 images
+	images, err := ds.imageV2DataStore.SearchRawImages(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert to list images
+	listImages := make([]*storage.ListImage, 0, len(images))
+	for _, image := range images {
+		// Convert v2 to v1
+		v1Image := mapper.ConvertToV1(image)
+		// Convert v1 to ListImage
+		listImage := types.ConvertImageToListImage(v1Image)
+		listImages = append(listImages, listImage)
+	}
+	return listImages, nil
+}
+
+func (ds *datastoreImpl) ListImage(ctx context.Context, sha string) (*storage.ListImage, bool, error) {
+	if !ds.flattenImageData {
+		return ds.imageDataStore.ListImage(ctx, sha)
+	}
+
+	// Get v2 image
+	image, found, err := ds.GetImageMetadata(ctx, sha)
```

Comment:
This might need the same logic as `GetImage` incase the given id is a SHA and not UUID.